docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,455 @@
+
+This is a link to the prototype for [ClusterAddonProvider for Helm](https://github.com/Jont828/cluster-api-addon-provider-helm), and a demo can be found in this [Cluster API Office Hours recording](https://www.youtube.com/watch?v=dipR1Tzh4jE). The prototype implements HelmChartProxy and HelmReleaseProxy as well as controllers for both CRDs.
+
+#### Defining the list of Cluster add-ons to be installed
+
+ClusterAddonProvider for Helm will use HelmChartProxy to define the addons to install and which clusters they should be installed on. Each HelmChartProxy will contain fields to specify a single Helm chart. Additionally, the spec will contain a key value pair which will be used as a cluster selector The Helm chart is then installed on every cluster that includes a label matching the cluster selector and makes it part of the orchestration. If a selected cluster no longer matches the cluster selector label, then the Helm release will be uninstalled. 
+
+Note that the cluster selector label differs from the match label field in ClusterResourceSets as HelmChartProxy specifies a single key value pair whereas the match labels field matches any number of labels. The match labels field was not used because by adding another label pair, all of the currently matched clusters would be unselected and that would trigger the deletion of all associated Helm releases. 
+
+This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.
+
+#### Install and lifecycle manage the Package Manager
+
+Some Package Manager might require in-cluster components, like e.g. the kapp-controller, however this is not the case for Helm starting from v3, so this is a no-op.
+In case this is required, the ClusterAddonProvider should take care of this task as well.
+In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.
```

Comment:
The HelmChartProxy controller imports Helm v3 as a library. From what I saw in [these docs](https://helm.sh/docs/topics/version_skew/#:~:text=As%20of%20Helm%203%2C%20Helm,n%2D1%20versions%20of%20Kubernetes), Helm 3 has limitations with which K8s versions it supports, but we can't really control that. So we'd need to pick the right version of Helm to use for an older K8s version.

Reply:
Okay and how does this related to "versions of HelmChartProxy"?

Reply:
I'm not exactly sure, I think was part of the original draft @fabriziopandini wrote, maybe he can comment about what he envisioned?

Reply:
If I remember well, this is just a reminder to document the support matrix of this new component if this does not overlap with CAPI support matrix

Reply:
Yup definitely something to keep in mind. On the other hand it might just be what they officially support (and test?) and it would also work with older/newer versions as long as they don't require new features. (somewhat like client-go also kind of works as long as you just use stable stuff)